pkg/ds/queue/queue.go

```diff
@@ -0,0 +1,155 @@
+package queue
+
+type Node[T any] struct {
+	value T
+	next  *Node[T]
+	prev  *Node[T]
+}
+
+type Queue[T any] struct {
+	head *Node[T]
+	tail *Node[T]
+	size uint
+}
+
+func New[T any]() *Queue[T] {
+	return &Queue[T]{
+		head: nil,
+		tail: nil,
+		size: 0,
+	}
+}
+
+func (q *Queue[T]) PushBack(val T) {
+	q.size++
+	node := &Node[T]{
+		value: val,
+		next:  nil,
+		prev:  nil,
+	}
+
+	if q.head == nil {
+		q.head = node
+		q.tail = node
+		return
+	}
+
+	node.prev = q.tail
+	q.tail.next = node
+	q.tail = node
+}
+
+func (q *Queue[T]) PushFront(val T) {
+	q.size++
+	node := &Node[T]{
+		value: val,
+		next:  nil,
+		prev:  nil,
+	}
+
+	if q.head == nil {
+		q.head = node
+		q.tail = node
+		return
+	}
+
+	node.next = q.head
+	q.head.prev = node
+	q.head = node
+}
+
+func (q *Queue[T]) PopBack() (T, bool) {
+	var t T
+
+	// 0 node
+	if q.tail == nil {
+		return t, false
+	}
+
+	q.size--
+	t = q.tail.value
+	// only 1 node
+	if q.tail == q.head {
```

Comment:
Check `q.Size` before `q.size--` makes code more readable.

Reply:
I have checked IsEmpty right before the `q.size--`. So adding 1 more check is unnecessary